pkg/controllermanager/controller/shoot/shoot_maintenance_control.go

```diff
@@ -378,6 +378,31 @@ func hasMaintainNowAnnotation(shoot *gardencorev1beta1.Shoot) bool {
 	return ok && operation == v1beta1constants.ShootOperationMaintain
 }
 
+func needsRetry(shoot *gardencorev1beta1.Shoot) bool {
+	needsRetryOperation := false
+
+	if val, ok := shoot.Annotations[v1beta1constants.FailedShootNeedsRetryOperation]; ok {
+		needsRetryOperation, _ = strconv.ParseBool(val)
+	}
+	delete(shoot.Annotations, v1beta1constants.FailedShootNeedsRetryOperation)
+
+	return needsRetryOperation
+}
+
+func getOperation(shoot *gardencorev1beta1.Shoot) string {
+	var (
+		operation            = v1beta1constants.GardenerOperationReconcile
+		maintenanceOperation = shoot.Annotations[v1beta1constants.GardenerMaintenanceOperation]
+	)
+
+	if isValid, _ := gardencorev1beta1helper.IsValidShootOperation(maintenanceOperation, shoot.Status.LastOperation); isValid {
```

Comment:
IIUC the second return value `removedBeforePersisted` is not used. Is it a leftover?

Reply:
Yeah, it's a left-over. I remove it for now. Initially, I intended to use this function also in `core/shoot/strategy.go`, but didn't do it for now.